numpyro/contrib/control_flow/scan.py

```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
                                f" but got {value_ndim}. Please report the issue to us!")
 
 
-def scan_wrapper(f, init, xs, length, reverse, rng_key=None, substitute_stack=[]):
+class promote_shapes(Messenger):
+    # a helper messenger to promote shapes of `fn` and `value`
```

Comment:
nit: add a comment or docstring with example old and new shapes for `fn` and `value` for maintainability.